From 33e2ebb88307569af3139c86f06286e004e37bf5 Mon Sep 17 00:00:00 2001
From: Lorenzo Rossi <git@rossilorenzo.dev>
Date: Fri, 10 Oct 2025 13:12:50 +0200
Subject: [PATCH] nvk: Fix QMD buffer length on upload

Current code allocates the maximum QMD data for all generations and
uploads everything, even on generations where a smaller QMD buffer
suffices. This is not only wasteful, but actually crashes Kepler GPUs
due to complications with the QMD queue.

Only upload the useful bytes of the QMD buffer.

Closes: https://gitlab.freedesktop.org/mesa/mesa/-/issues/14070
Fixes: 0e268dad00c ("nvk: Allow for larger QMDs")
Signed-off-by: Lorenzo Rossi <git@rossilorenzo.dev>
---
 src/nouveau/vulkan/nvk_cmd_dispatch.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/nouveau/vulkan/nvk_cmd_dispatch.c b/src/nouveau/vulkan/nvk_cmd_dispatch.c
index 2d8c9ced06fbc..afd879a099e73 100644
--- a/src/nouveau/vulkan/nvk_cmd_dispatch.c
+++ b/src/nouveau/vulkan/nvk_cmd_dispatch.c
@@ -228,12 +228,12 @@ nvk_cmd_upload_qmd(struct nvk_cmd_buffer *cmd,
       nak_fill_qmd(&pdev->info, &shader->info, &qmd_info, qmd, qmd_size_B);
 
       void *qmd_map;
-      result = nvk_cmd_buffer_alloc_qmd(cmd, sizeof(qmd), 0x100,
+      result = nvk_cmd_buffer_alloc_qmd(cmd, qmd_size_B, 0x100,
                                         &qmd_addr, &qmd_map);
       if (unlikely(result != VK_SUCCESS))
          return result;
 
-      memcpy(qmd_map, qmd, sizeof(qmd));
+      memcpy(qmd_map, qmd, qmd_size_B);
    }
 
    *qmd_addr_out = qmd_addr;
-- 
GitLab

